feat(bootstrap): split Bootstrap::new method into two

RolandSherwin opened the pull request:
- Split the `async fn Bootstrap::new` method into two methods, where the `new` is made a "sync" method and `async fn new_with_preloaded_addrs` retains the same functionality as the old new method.
- The `new_with_preloaded_addrs` method is only required inside clients which tries to pre-populate its RT before starting the libp2p driver.
- Introduces `is_addr_queue_empty`, as it will be used by the reachability check implementation
- Deprecate `try_next_dial_addr` as it is just a copy of `next_addr`

RolandSherwin committed:
feat(bootstrap)!: split Bootstrap::new method into two for easy
integration
- Split the `async fn Bootstrap::new` method into two methods, where the `new`
  is made a "sync" method and `async fn new_with_preloaded_addrs` retains the
  same functionality as the old new method.
- The `new_with_preloaded_addrs` method is only required inside clients which
  tries to pre-populate its RT before starting the libp2p driver.
- Introduces `is_addr_queue_empty`, as it will be used by the reachability
  check implementation
- Deprecate `try_next_dial_addr` as it is just a copy of `next_addr`

RolandSherwin committed:
feat(bootstrap): revamp the tests to make it much more detailed

maqi commented on ant-bootstrap/src/bootstrap.rs:
```diff
@@ -79,7 +79,11 @@ pub struct Bootstrap {
+    /// Use `new_with_preloaded_addrs` to ensure that the struct contains at least MIN_INITIAL_ADDRS addresses immediately.
+    ///
+    /// Must be called from a tokio runtime context.
+    pub fn new(config: BootstrapConfig) -> Result<Self> {
```
> to ensure that the struct contains at least MIN_INITIAL_ADDRS addresses immediately

Does this need to be `always enforced`, i.e. replace this `new` with `new_with_preloaded_addrs`, or raise an error if not having MIN_INITIAL_ADDRS addresses ?

RolandSherwin replied:
Updated the comments to reflect the changes!

maqi commented on ant-bootstrap/src/lib.rs:
```diff
@@ -62,49 +62,30 @@ pub fn craft_valid_multiaddr(addr: &Multiaddr, ignore_peer_id: bool) -> Option<M
+    output_address.push(udp);
+    let quic = addr
         .iter()
-        .find(|protocol| matches!(protocol, Protocol::Tcp(_)));
```
just to confirm that with this removal, TCP/WebSocket transport is no longer supported.

RolandSherwin replied:
Yes, TCP/ws are not supported in current main. Just udp, this is just cleaning up the bootstrap method accordingly.

maqi commented on ant-bootstrap/src/bootstrap.rs:
```diff
@@ -135,11 +139,22 @@ impl Bootstrap {
-            info!("First node in network; clearing any existing cache");
-            bootstrap.cache_store.write().await?;
-            return Ok(bootstrap);
+            tokio::spawn(async move {
```
* The spawned task has a clone of cache_store, but the original bootstrap uses a different clone - is this intentional?
* A first node that fails to clear its cache will continue running with stale data

Bit edge cased, but may happen ?

RolandSherwin replied:
The struct uses `Arc` internally, so we should be fine. Yeah, the other thing is an edge case, but we usually run the `first node` on our machines, so ig we were fine earlier and we should be fine now

maqi commented on ant-node/src/bin/antnode/main.rs:
```diff
@@ -317,7 +317,7 @@ fn main() -> Result<()> {
     let bootstrap_config = BootstrapConfig::try_from(&opt.peers)?
         .with_backwards_compatible_writes(opt.write_older_cache_files);
-    let bootstrap = rt.block_on(Bootstrap::new(bootstrap_config))?;
+    let bootstrap = rt.block_on(async { Bootstrap::new(bootstrap_config) })?;
```
Since `Bootstrap::new()` is now sync, wrapping it in `async {}` and block_on is unnecessary. This should just be:
```
let bootstrap = Bootstrap::new(bootstrap_config)?;
```

RolandSherwin replied:
Nice catch, updated it!

maqi commented on ant-bootstrap/src/bootstrap.rs:
```diff
@@ -250,6 +265,29 @@ impl Bootstrap {
+    /// Check if the address queue is empty without consuming an address.
+    /// Returns `Some(true)` if empty, `Some(false)` if not empty, and `None` if still fetching or has cache or contacts
+    /// sources pending.
+    pub fn is_addr_queue_empty(&self) -> Option<bool> {
```
This three-state return value could be confusing. Consider using an enum for clarity:
```
enum QueueStatus { HasAddresses, Empty, FetchPending }
```

RolandSherwin replied:
`is_addr_queue_empty` is used only inside the tests for now, so we should be good?

maqi commented on ant-bootstrap/src/bootstrap.rs:
```diff
@@ -135,11 +139,22 @@ impl Bootstrap {
+    }
+
+    /// Create a new Bootstrap manager and ensure it has at least MIN_INITIAL_ADDRS addresses preloaded.
+    pub async fn new_with_preloaded_addrs(config: BootstrapConfig) -> Result<Self> {
```
It seems it's now preferred to always use `new_with_preloaded_addrs` ?
i.e. changing `Bootstrap::new(config)` to `Bootstrap::new_with_preloaded_addrs(config).await` ?
if so, this will be a breaking change requires all downstream callers to be updated ?

RolandSherwin replied:
Yes, marking the commit as a breaking one

RolandSherwin committed:
docs(bootstrap): clarify new vs new_with_preloaded_addrs usage

- Update Bootstrap::new() docs to explain lazy address fetching
- Update Bootstrap::new_with_preloaded_addrs() docs to explain preloading

RolandSherwin committed:
refactor(bootstrap): make first node cache clear synchronous

Replace async cache_store.write() call with synchronous file deletion
using std::fs::remove_file(). This makes Bootstrap::new() truly sync
as documented, removing the spawned async task for cache clearing.
